### PR TITLE
Added post about performance of the Versioned HDF5 library.

### PR DESCRIPTION
**Pull request by melissawm**
I think this still needs a review, especially the images (sizing and position).

### CODE REVIEW

**Comment by melissawm** [comment]
I think @scopatz should take a look too, if he can.

File: posts/2020/08/versioned-hdf5-performance.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+.. date: 2020-08-31 08:00:00 UTC-05:00
+.. author: Melissa Weber Mendonça
+.. tags: HDF5, h5py
+.. category:
+.. link:
+.. description:
+.. type: text
+-->
+
+In a previous post, we introduced the Versioned HDF5 library and described some of its features. In this post, we'll show some of the performance analysis we did while developing the library, hopefully making the case that reading and writing versioned HDF5 files is efficient and useful.
```
Comment (scopatz) [inline posts/2020/08/versioned-hdf5-performance.md:13]: Link to previous post needed, probably

File: posts/2020/08/versioned-hdf5-performance.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+We’ll start by analyzing how the HDF5 file sizes grow as the number of versions grows. Using a chunk size of **4096**, we can see the following results for the 4 one-dimensional test cases:
+
+<img src="/images/versioned-hdf5-performance/filesizes.png" alt="File sizes for Versioned HDF5 files" width="533" height="400">
+
+Note that in some of the examples the size of the arrays stored in the file also grow as the number of versions grows, since each new version is changing the original arrays by adding, deleting and changing values in the original arrays. *Keep in mind there is redundant data as some of it is not changed during the staging of a new version but it is still being stored.* It's also worth noting that for these tests we don't use compression, even though algorithms available in [h5py](link to h5py compression) can be used in Versioned HDF5. You can see the [Versioned HDF5 documentation](https://deshaw.github.io/versioned-hdf5/) for more detailed tests regarding chunking and compression.
```
Comment (scopatz) [inline posts/2020/08/versioned-hdf5-performance.md:56]: `[h5py](link to h5py compression)` should be a real link :wink:

File: posts/2020/08/versioned-hdf5-performance.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+![](/images/versioned-hdf5-performance/latestread.png)
+
+In this case, we can see that on average, reading the latest version on a VersionedHDF5File is ~5x-10x slower than reading an unversioned file. Also, the time required to read the latest version from a Versioned HDF5 file increases modestly with the number of versions stored in the file, except when the size of the array increases significantly with each new version.
+
+## Summary
+
+Overall, the worst performance was observed for tests with large array sizes. This seems to show that the file sizes and IO performance of Versioned HDF5 files are significantly affected by the size of the datasets stored in each file, which is to be expected.
```
Comment (scopatz) [inline posts/2020/08/versioned-hdf5-performance.md:106]: Probably the size of `unique` data

**Comment by scopatz** [comment]
Thanks so much for putting this in!

**Comment by asmeurer** [comment]
Unless this is merged today, this will need to be moved to posts/2020/09/

**Comment by rgommers** [comment]
> Unless this is merged today, this will need to be moved to posts/2020/09/

That's fine, I can do that - I'm in the habit of updating the date/time anyway, so it's as recent as possible. That helps posts show up at the top of Planet SciPy / Planet Python.

**Comment by melissawm** [comment]
Links fixed, let me know if there's anything else. Otherwise I'm happy with it!

File: posts/2020/08/versioned-hdf5-performance.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+In a [previous post](https://labs.quansight.org/blog/2020/08/introducing-versioned-hdf5/), we introduced the Versioned HDF5 library and described some of its features. In this post, we'll show some of the performance analysis we did while developing the library, hopefully making the case that reading and writing versioned HDF5 files is efficient and useful.
+
+<!-- TEASER_END -->
+
+## What are we measuring?
+
+*Performance* can mean different things for different operations. For the tests presented here, the main goals are:
+
+- To evaluate the performance of reading/writing Versioned HDF5 files and compare it with non-versioned files (that is, files where only the last version of the datasets is stored);
```
Comment (rgommers) [inline posts/2020/08/versioned-hdf5-performance.md:21]: _"*Performance* can mean different things" ... "To evaluate the performance "_ is a little odd - maybe add (size on disk, and I/O speed) in brackets?

File: posts/2020/08/versioned-hdf5-performance.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+This is the second post in a series about the Versioned HDF5 library. Next,
+we'll discuss the the design of the Versioned HDF5 library.
+
+`versioned-hdf5 1.0` has recently been released, and is available on PyPI and conda-forge. You can install it with
+
+```py
+conda install -c conda-forge versioned-hdf5
+```
+
+The development is on [GitHub](https://github.com/deshaw/versioned-hdf5).
```
Comment (rgommers) [inline posts/2020/08/versioned-hdf5-performance.md:122]: I propose a slight rephrase: "Development for `versioned-hdf5` happens on [GitHub]"

File: posts/2020/08/versioned-hdf5-performance.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+Next steps
+----------
+
+This is the second post in a series about the Versioned HDF5 library. Next,
+we'll discuss the the design of the Versioned HDF5 library.
+
+`versioned-hdf5 1.0` has recently been released, and is available on PyPI and conda-forge. You can install it with
+
+```py
```
Comment (rgommers) [inline posts/2020/08/versioned-hdf5-performance.md:118]: `bash` not `py` I'd think

File: posts/2020/08/versioned-hdf5-performance.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+## Summary
+
+Overall, the worst performance was observed for tests with large array sizes. This seems to show that the file sizes and IO performance of Versioned HDF5 files are significantly affected by the size of the unique datasets stored in each file, which is to be expected. 
+
+The results presented here show that the largest impact on IO performance and storage is in fact explained by the size of the datasets stored in the file, and not significantly reduced by the Versioned HDF5 abstraction.
+
+Next steps
+----------
+
+This is the second post in a series about the Versioned HDF5 library. Next,
```
Comment (rgommers) [inline posts/2020/08/versioned-hdf5-performance.md:113]: textual: "next" -> "In our next post"

File: posts/2020/08/versioned-hdf5-performance.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+![](/images/versioned-hdf5-performance/latestread.png)
+
+In this case, we can see that on average, reading the latest version on a VersionedHDF5File is ~5x-10x slower than reading an unversioned file. Also, the time required to read the latest version from a Versioned HDF5 file increases modestly with the number of versions stored in the file, except when the size of the array increases significantly with each new version.
+
+## Summary
+
+Overall, the worst performance was observed for tests with large array sizes. This seems to show that the file sizes and IO performance of Versioned HDF5 files are significantly affected by the size of the unique datasets stored in each file, which is to be expected. 
+
+The results presented here show that the largest impact on IO performance and storage is in fact explained by the size of the datasets stored in the file, and not significantly reduced by the Versioned HDF5 abstraction.
```
Comment (rgommers) [inline posts/2020/08/versioned-hdf5-performance.md:108]: IO -> I/O

File: posts/2020/08/versioned-hdf5-performance.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+As expected, read times increase for files with a larger number of versions, but the growth is close to linear in all cases except for `test_mostly_appends_dense`, where the large array sizes explain the larger read times.
+
+Next, we’ll compare the times necessary to read the latest version on each file. Because of how VersionedHDF5 is designed, this is the same as reading the last array stored in the HDF5 file. For each test, we made 20 different reads of the latest version in order to eliminate fluctuations generated by background processes or the OS (solid lines). We also compare these read times with the time required to read an unversioned file (which only stored the latest version - dashed black line).
+
+![](/images/versioned-hdf5-performance/latestread.png)
+
+In this case, we can see that on average, reading the latest version on a VersionedHDF5File is ~5x-10x slower than reading an unversioned file. Also, the time required to read the latest version from a Versioned HDF5 file increases modestly with the number of versions stored in the file, except when the size of the array increases significantly with each new version.
```
Comment (rgommers) [inline posts/2020/08/versioned-hdf5-performance.md:102]: `VersionedHDF5File` should be typeset as code

File: posts/2020/08/versioned-hdf5-performance.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+   
+In the next tests, the number of modifications is dominated by the number of appended rows. There are two such cases:
+
+4. In the first case, the dataset contains three one-dimensional arrays with 1000 rows initially, and 1000 rows are added with each new version. A small number (at most 10) values are chosen at random, following the power law described above, and changed or deleted. We call this test `test_mostly_appends_sparse`.
+
+5. In the second case, the dataset contains one two-dimensional array with shape `(30, 30)` and two one-dimensional arrays acting as indices to the 2d array. In this case, rows are only appended in the first axis of the two-dimensional array, and a small number of positions (at most 10) is chosen at random and changed. We call this test `test_mostly_appends_dense`.
+
+To test the performance of VersionedHDF5 files, we have chosen to compare a few different chunk sizes and compression algorithms. These values have been chosen arbitrarily, and optimal values depend on different use cases and on the nature of the datasets stored in the file.
```
Comment (rgommers) [inline posts/2020/08/versioned-hdf5-performance.md:46]: You're using "Versioned HDF5" for the library name (e.g. in blog title), so should this be typeset as code maybe? Or have a space?

Same applies to `VersionedHDF5` in other paragraphs

**Review by rgommers — COMMENTED** [review]
Thanks Melissa. This post reads really well, I have only a few small comments.

File: posts/2020/08/versioned-hdf5-performance.md (hunk 10 above)
Comment (melissawm) [reply posts/2020/08/versioned-hdf5-performance.md:46]: Thanks for that. I made a pass and tried to be consistent and use "Versioned HDF5" for the library, and "versioned HDF5 files" for the actual resulting files. Let me know if this is still confusing, and I'll rephrase.

**Comment by melissawm** [comment]
I updated the post to reflect some comments from DE Shaw. Let me know if there's anything else!

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks Melissa. This is looking really polished. There is one significant issue though: all the conclusions are too negative, and the positive news is left unstated! In your introduction you say you want to convince the reader that using Versioned HDF5 is efficient and useful - but you're not doing that. A careful reader may draw their own conclusions from studying the figures, but that's a big ask of a reader.

For example, in the summary your two paragraphs start with: _Overall, the worst performance was observed ..._ and _The results presented here show that the largest impact on ..._

Here is what I suggest:
- Under the "File sizes" figure, write something like "We observe that writing a thousand versions of a 3.3 MB array - something that would take 3.3 GB if stored as separate files - takes about 77 MB, a storage size reduction of 40x!"
- Find a good place for a more explicit conclusion about creation times and I/O performance. E.g. when you write _"increases significantly when compared to regular HDF5 files. This is expected ..."_ give a number at the end of the paragraph, like "increases linearly, which is our desired behavior, and the creation time cost is only 4x for on-disk storage savings of 40x!"
- And then start with the positive news in the summary.

**Comment by melissawm** [comment]
Thanks @rgommers , you are totally right, I think that was missing from my perspective because of looking at the results for so long! I've added some comments to try to make those ideas clearer.

File: posts/2020/09/versioned-hdf5-performance.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+<!-- TEASER_END -->
+
+## What are we measuring?
+
+*Performance* can mean different things for different operations. For the tests presented here, the main goals are:
+
+- To evaluate the performance (size on disk and I/O speed) of reading/writing versioned HDF5 files and compare it with non-versioned files (that is, files where only the last version of the datasets is stored);
+- To evaluate the performance when reading/writing data to a versioned HDF5 file in a set of different use cases;
+- To evaluate the performance when different parameters options are considered for chunking and compression on versioned HDF5 files.
```
Comment (scopatz) [inline posts/2020/09/versioned-hdf5-performance.md:23]: This should either be `parameter's options` (plural possessive) or `parameter options` (plural)

File: posts/2020/09/versioned-hdf5-performance.md [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+When different versions of a dataset are stored in a versioned HDF5 file, modified copies of the data are stored as new versions. This means that there may be duplication of data between versions, which might impact the performance of reading, writing or manipulating these files.
+
+In order to analyze this, we will consider test files created with a variable number of versions (or transactions) of a dataset consisting of three ndarrays of variable length. One test includes a two-dimensional ndarray as part of this dataset, but all other test cases consist of three one-dimensional ndarrays per dataset.
+
+With each new version a certain number of rows is added, removed, and modified. For
```
Comment (scopatz) [inline posts/2020/09/versioned-hdf5-performance.md:29]: `rows are added`

File: posts/2020/09/versioned-hdf5-performance.md [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+these tests, all changes are made to elements chosen according to a [power law](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.powerlaw.html) which biases the modifications towards the end of the array, simulating a possible use case of modifying more recent results in a given timeseries.
+
+The tests are as follows:
+
+1. A large fraction of changes is made to the dataset with each new version: The dataset initially has three arrays with 5000 rows, and 1000 positions are chosen at random and changed, and a small number (at most 10) rows are added or deleted with each new version. We will refer to this test as `test_large_fraction_changes_sparse`.
+
+2. A small fraction of changes is made to the dataset with each new version: The dataset initially has three arrays with 5000, but only 10 positions are chosen at random and changed, and a small number (at most 10) rows are added or deleted with each new version. We will refer to this test as `test_small_fraction_changes_sparse`.
```
Comment (scopatz) [inline posts/2020/09/versioned-hdf5-performance.md:36]: `with 5000 rows`

**Review by scopatz — APPROVED** [review]
Just a couple of small suggestions from me!

**Comment by melissawm** [comment]
Thanks @scopatz !

**Comment by rgommers** [comment]
Hey @melissawm, @asmeurer, @scopatz. I re-read both the first and second versioned-hdf5 blog posts, as well as the README on the repo and the docs. It all kind of looks as expected, but there's still no "here's why you should be excited" selling points that jump out at me. I think this is actually important, especially with the 1.0 release coming up.

My first take-away from some graphs was a 40x size-on-disk reduction, but after Melissa's update I see it's only 50%. 

So is the reason someone should care more like:
- Some data naturally fits HDF5 more than, say, a database
- If that data changes over time, it's a headache to keep many versions manually, both in terms of storing many files and in manually keeping track of versions in for example an index file.
- Hence we'd like an efficient way to keep many versions inside of one HDF5 file.
- Blog post 1 shows this works, with a nice intuitive API.
- Blog post 2 shows that this also saves on disk space - which is another important benefit -  and further reductions in this area can be achieved in the future with use of hdf5/h5py-provided compression algorithms.
- That only comes at a cost of <10x file writing speed; file reading speed for the latest version of the data is unaffected.

Does that about capture it?

**Comment by scopatz** [comment]
Yeah - the only thing that I would add is that the motivation for this is to have any mechanism of storing binary data sets in a versioned way, that feels natural to users of other version control systems. Of course, we want the implementation to be *as efficient as possible* (which is what Ralf outlined, and what these blog posts prove).  The novelty here is that it is a standard implementation of these version control ideas for binary data and data that has a numpy-like model. 

@melissawm - Can you find some time to squeeze of the Ralf's text and this motivation in?

**Comment by melissawm** [comment]
Done, let me know if this is better. I reworked the intro and summary at the end.

**Comment by scopatz** [comment]
I think some of the summary contents should be copied / moved to the intro at the top.  We kind of want a TL;DR / abstract here. Thanks @melissawm!

**Comment by scopatz** [comment]
LGTM now! @rgommers?

**Review by rgommers — APPROVED** [review]
Okay, in it goes! Great work @melissawm & all!

**Comment by melissawm** [comment]
Thanks for all the helpful comments!

**Pull request merged**